src/NitroSharp/Content/ArchiveManager.cs

```diff
@@ -0,0 +1,255 @@
+                }
+                catch (ArgumentException)
+                {
+                    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
+                    encoding = Encoding.GetEncoding(DefaultEncoding);
+                }
+            }
+            _encoding = encoding;
+            _root = new VFSNode();
+        }
+
+        ~ArchiveManager()
+        {
+            Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    _root?.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public void AddMounts(MountPoint[] mountPoints, string rootDirectory, ContentManager content)
+        {
+            foreach (MountPoint mountPoint in mountPoints)
+            {
+                AddMount(mountPoint, rootDirectory, content);
+            }
+        }
+
+        private void AddMount(MountPoint mountPoint, string rootDirectory, ContentManager content)
+        {
+            string[] splitedPath = mountPoint.MountName.ToLowerInvariant().Split("/");
+
+            VFSNode head = _root;
+            foreach (string part in splitedPath)
+            {
+                if (head.Children == null)
+                {
+                    head.Children = new Dictionary<string, VFSNode>();
+                }
+
+                if (!head.Children.ContainsKey(part))
+                {
+                    head.Children[part] = new VFSNode();
+                }
+
+                head = head.Children[part];
+            }
+
+            if (head.MountedArchives == null)
+            {
+                head.MountedArchives = new List<IArchiveFile>();
+            }
+
+            string fullPath = Path.Combine(rootDirectory, mountPoint.ArchiveName);
+            if (File.Exists(fullPath))
+            {
+                Func<MemoryMappedFile,Encoding,IArchiveFile>[] constructors = new Func<MemoryMappedFile,Encoding,IArchiveFile>[]
+                {
+                    NPAFile.Create,
+                    AFSFile.Create,
+                };
+                MemoryMappedFile mmFile = MemoryMappedFile.CreateFromFile(fullPath, FileMode.Open, null, 0, MemoryMappedFileAccess.Read);
+                IArchiveFile? archive = null;
+                foreach (Func<MemoryMappedFile,Encoding,IArchiveFile> constructor in constructors)
+                {
+                    try
+                    {
+                        archive = constructor(mmFile, _encoding);
+                        archive.OpenArchive();
+                        break;
+                    }
+                    catch
+                    {
+                        continue;
+                    }
+                }
```

Comment:
This seems overly generic to me. We know there's only gonna be 2 archive formats, so is there really a need for this much indirection? Not a fan of using exceptions for this either — you're actually expecting the load operation to fail, there's nothing "exceptional" about it. Consider introducing a static method called `TryLoad` instead.

...Hold on, why can't we just look at the file extension?

Reply:
This is defenitly personal preference but I just prefer to trust magic bytes instead of the extension. For example, the subarchives of `voice` do not have `.afs` in their names. A cleaner way might be to map magic bytes to the classes instead of just using a list but since there is only 2 formats right now, I think it's the cleanest way to do it.

I also chose this structure because it's the most flexible, while noah only uses AFS and NPA I have no idea if other N2 games support other formats. I know the goal right now is just Xbox 360 Noah, but just this small future-proofing doesn't seem overkill.

Reply:
If you're gonna go down this route, you should probably go all the way 🙂 
```csharp
IArchiveFile? archive = loadMethods
    .Select(load => load(mmFile, _encoding))
    .FirstOrDefault(x => x is not null);
```